enhancements/api-review/gcp_user_defined_labels.md

```diff
@@ -0,0 +1,305 @@
+resource even if the labels are removed from this field(despite it being immutable).
+
+If the userLabels field is changed post-install, there is no guarantee about how an 
+in-cluster operator will respond to the change. Some operators may reconcile the 
+change and update the labels on the GCP resource. Some operators may ignore the change. 
+However, if labels are removed from userLabels, the label will not be removed from the 
+GCP resource.
+
+### Workflow Description
+
+- An Openshift administrator requests to add required labels to all GCP resources 
+  created by Openshift by adding it in `platform.gcp.userLabels`
+- Openshift installer validates the labels defined in `platform.gcp.userLabels` and 
+  adds these labels to all resources created during installation and also updates 
+  `.status.platformStatus.gcp.resourceLabels` of the `infrastructure.config.openshift.io`
+- In cluster operators refers `.status.platformStatus.gcp.resourceLabels` of the 
+  `infrastructure.config.openshift.io` to add labels to the resources created later.
+
+#### Variation [optional]
+
+### API Extensions
+Enhancement requires below modifications to the mentioned CRDs
+- Add `userLabels` field to `platform.gcp` of the `installconfigs.install.openshift.io`
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: installconfigs.install.openshift.io
+spec:
+  versions:
+  - name: v1
+    schema:
+      openAPIV3Schema:
+        properties:
+          platform:
+            properties:
+              gcp:
+                properties:
+                  userLabels:
+                    additionalProperties:
+                      type: string
+                    description: UserLabels additional keys and values that the installer
+                      will add as labels to all resources that it creates. Resources
+                      created by the cluster itself may not include these labels.
+                  type: object
+```
+
+- Add `resourceLabels` field to `status.platformStatus.gcp` 
+  of the `infrastructure.config.openshift.io`
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: infrastructures.config.openshift.io
+spec:
+  versions:
+  - name: v1
+    schema:
+      openAPIV3Schema:
+        properties:
+          status:
+            properties:
+              platformStatus:
+                properties:
+                  gcp:
+                    properties:
+                      resourceLabels:
+                        description: resourceLabels is a list of additional labels to apply to GCP resources created for the cluster. See https://cloud.google.com/compute/docs/labeling-resources for information on labeling GCP resources. GCP supports a maximum of 64 labels per resource. OpenShift reserves 14 labels for its internal use, and allows 50 labels for user configuration.
+                        type: array
+                        maxItems: 50
```

Comment:
how was this limit determined?

Reply:
I thought since customer asked for support up to 40 tags in AWS, keeping max for user configuration would be ideal.

Reply:
To be aligned with the AWS implementation, updated the number of tags supported to 32 for user configuration and 32 for OCP's internal use, shared equally between the two.